clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java

```diff
@@ -303,21 +386,42 @@ private void completeBatch(RecordBatch batch, ProduceResponse.PartitionResponse
         if (error != Errors.NONE && canRetry(batch, error)) {
             // retry
             log.warn("Got error produce response with correlation id {} on topic-partition {}, retrying ({} attempts left). Error: {}",
-                     correlationId,
-                     batch.topicPartition,
-                     this.retries - batch.attempts - 1,
-                     error);
-            this.accumulator.reenqueue(batch, now);
-            this.sensors.recordRetries(batch.topicPartition.topic(), batch.recordCount);
+                    correlationId,
+                    batch.topicPartition,
+                    this.retries - batch.attempts - 1,
+                    error);
+            if (transactionState == null || (transactionState.pidAndEpoch().pid == batch.pid())) {
+                // If idempotence is enabled only retry the request if the current PID is the same as the pid of the
+                // batch.
+                if (transactionState != null) {
+                    log.debug("Retrying batch to {}-{}. Sequence number : {}", batch.topicPartition.topic(), batch.topicPartition.partition(), transactionState.sequenceNumber(batch.topicPartition));
+                }
+                this.accumulator.reenqueue(batch, now);
+                this.sensors.recordRetries(batch.topicPartition.topic(), batch.recordCount);
+            } else {
+                // TODO(apurva): Is 'KafkaException' the best one to throw here? What would be more appropriate?
```

Comment:
Maybe we should treat this as an `OutOfSequenceException`. That is the error which caused us to get a new pid and probably what would happen if we tried to resend this batch.